File: install-page/determine.py
"""Orbit determination: IOD + differential correction + outlier rejection.

Reproducer for the "Determine an Orbit" section of
https://empyrean-dynamics.com/install. Reads ADES PSV from a sample file in
this directory, fits an orbit, then inspects and refines that same fit —
each step hands its ``orbit`` to the next.
"""

from __future__ import annotations

from pathlib import Path

import empyrean

PSV = str(Path(__file__).parent / "observations.psv")


def main() -> None:
    empyrean.initialize()

    # spielberg:snippet:start
    # Read ADES PSV observations (file path or PSV string)
    obs = empyrean.read_ades(PSV)

    # Full pipeline: IOD + differential correction + outlier rejection
    fit = empyrean.determine(obs)
    print(
        f"converged={fit.converged}  "
        f'RMS RA·cos(d) {fit.summary.rms_ra_arcsec:.2f}" '
        f'Dec {fit.summary.rms_dec_arcsec:.2f}"'
    )

    # `fit.orbit` is a re-feedable orbit (state + covariance + non-grav).
    # Evaluate its own residuals against the arc — no fitting:
    ev = empyrean.evaluate(fit.orbit, obs)
    print(f"post-fit chi2 = {ev.summary.chi2:.1f}")

    # Refine the fit with a Bayesian prior (the fitted covariance is the
    # prior). The refined result is itself re-feedable into the next refine.
    refined = empyrean.refine(fit.orbit, obs)
    print(f"refined: converged={refined.converged}")
    # spielberg:snippet:end
    _ = refined


if __name__ == "__main__":
    main()

File: install-page/determine.rs
//! Orbit determination: IOD + differential correction + outlier rejection.
//!
//! Reproducer for the "Determine an Orbit" section of
//! <https://empyrean-dynamics.com/install>. Reads ADES PSV from a sample
//! file in this directory, fits an orbit, then inspects and refines that
//! same fit — each step hands its `orbit` to the next.

use empyrean::Context;

fn main() -> empyrean::Result<()> {
    let ctx = Context::from_data_dir(None)?;

    let psv_path = concat!(env!("CARGO_MANIFEST_DIR"), "/install-page/observations.psv");
    let psv = std::fs::read_to_string(psv_path).expect("sample observations.psv not found");

    // empyrean:snippet:start
    use empyrean::ODConfig;

    let cfg = ODConfig::default();

    // Read ADES PSV observations (file path or PSV string). The returned
    // set carries both ADES tables (optical + radar); the Python twin
    // unpacks the (optical, radar) tuple and folds the radar back in.
    let obs = ctx.read_ades(&psv)?;

    // Full pipeline: IOD + differential correction + outlier rejection.
    // Any <radar> block read above rides along inside `obs`.
    let fit = ctx.determine(&obs, None, &cfg)?.into_single()?;
    let s = &fit.summary;
    println!(
        "converged={}  RMS RA\u{00B7}cos(d) {:.2}\"  Dec {:.2}\"",
        fit.converged, s.rms_ra_arcsec, s.rms_dec_arcsec
    );

    // `fit.orbit` is a re-feedable Orbit (state + covariance + non-grav).
    // Evaluate its own residuals against the arc — no fitting:
    let eval = ctx.evaluate(&fit.orbit, &obs, &cfg)?;
    println!("post-fit chi2 = {:.1}", eval.summary.chi2);

    // Refine the fit with a Bayesian prior (the fitted covariance is the
    // prior). The refined result is itself re-feedable into the next refine.
    let refined = ctx.refine(&fit.orbit, &obs, &cfg)?;
    println!("refined: converged={}", refined.converged);
    // empyrean:snippet:end
    let _ = refined;
    Ok(())
}
